src/channel/ws_0_8.ts

```diff
@@ -0,0 +1,564 @@
+   * @returns subscriptionId | onerror(Event)
+   * @example
+   * ```typescript
+   * const subscriptionId = await webSocketChannel.waitForUnsubscription();
+   * ```
+   */
+  public async waitForUnsubscription(forSubscriptionId?: number) {
+    // unsubscribe
+    return new Promise((resolve, reject) => {
+      if (!this.websocket) return;
+      this.onUnsubscribeLocal = (subscriptionId) => {
+        if (forSubscriptionId === undefined) {
+          resolve(subscriptionId);
+        } else if (subscriptionId === forSubscriptionId) {
+          resolve(subscriptionId);
+        }
+      };
+      this.websocket.onerror = reject;
+    });
+  }
+
+  /**
+   * Reconnect re-create this.websocket instance
+   */
+  public reconnect() {
+    this.websocket = new WebSocket(this.nodeUrl);
+
+    this.websocket.addEventListener('open', this.onOpen.bind(this));
+    this.websocket.addEventListener('close', this.onCloseProxy.bind(this));
+    this.websocket.addEventListener('message', this.onMessageProxy.bind(this));
+    this.websocket.addEventListener('error', this.onError.bind(this));
+  }
+
+  // TODO: Add/Test ping service. It seems this work out of the box from pathfinder. If net disc. it will auto replay.
+  private reconnectAndUpdate() {
+    this.reconnect();
+    // TODO: attempt n reconnection times
+    // TODO: replay data from last block received (including it) up to latest
+  }
+
+  private onCloseProxy(ev: CloseEvent) {
+    this.websocket.removeEventListener('open', this.onOpen);
+    this.websocket.removeEventListener('close', this.onCloseProxy);
+    this.websocket.removeEventListener('message', this.onMessageProxy);
+    this.websocket.removeEventListener('error', this.onError);
+
+    this.onClose(ev);
+  }
+
+  private onMessageProxy(event: MessageEvent<any>) {
+    const message: WebSocketEvent = JSON.parse(event.data);
+    // console.log('onMessage:', data);
+    switch (message.method) {
+      case 'starknet_subscriptionReorg':
+        throw Error('Reorg'); // todo: implement what to do
+      case 'starknet_subscriptionNewHeads':
+        this.onsNewHeads(message.params as SubscriptionNewHeadsResponse);
+        break;
+      case 'starknet_subscriptionEvents':
+        this.onEvents(message.params as SubscriptionEventsResponse);
+        break;
+      case 'starknet_subscriptionTransactionsStatus':
+        this.onTransactionStatus(message.params as SubscriptionTransactionsStatusResponse);
+        break;
+      case 'starknet_subscriptionPendingTransactions':
+        this.onPendingTransaction(message.params as SubscriptionPendingTransactionsResponse);
+        break;
+      default:
+        break;
+    }
+    this.onMessage(event);
+  }
+
+  /**
+   * subscribe to new block heads
+   * * you can subscribe to this event multiple times and you need to manage subscriptions manually
+   */
+  public subscribeNewHeadsUnmanaged(blockIdentifier?: BlockIdentifier) {
```

Comment:
maybe to group all public methods in one place and the private ones on the bottom? now we have some public, and then private and later again public methods

Reply:
Yea but they are grouped by logical semantic. Feel free to reorder :D